src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,385 @@
+            mount->base_path, &host_config, &slot_config, &mount_config, &mount->handle.card);
+#endif
+
+    } else if (!strcmp(source, "sdspi")) {
+        mount_config.allocation_unit_size = 512;
+
+        sdmmc_host_t host_config = SDSPI_HOST_DEFAULT();
+        sdspi_device_config_t spi_slot_config = SDSPI_DEVICE_CONFIG_DEFAULT();
+
+        term spi_port = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x8", "spi_host"), term_invalid_term(), ctx->global);
+        spi_host_device_t host_dev;
+        // spi_driver_get_peripheral already checks if spi_port is valid
+        bool ok = spi_driver_get_peripheral(spi_port, &host_dev, ctx->global);
+        if (!ok) {
+            free(source);
+            free(target);
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        spi_slot_config.host_id = host_dev;
+
+        term cs_term = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x2", "cs"), term_invalid_term(), ctx->global);
+        if (UNLIKELY(!term_is_integer(cs_term))) {
+            free(source);
+            free(target);
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        spi_slot_config.gpio_cs = term_to_int(cs_term);
+
+        term cd_term = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x2", "cd"), UNDEFINED_ATOM, ctx->global);
+        if (cd_term != UNDEFINED_ATOM) {
+            if (UNLIKELY(!term_is_integer(cd_term))) {
+                free(source);
+                free(target);
+                RAISE_ERROR(BADARG_ATOM);
+            }
+            spi_slot_config.gpio_cd = term_to_int(cd_term);
+        }
+
+        mount = enif_alloc_resource(platform->mounted_fs_resource_type, sizeof(struct MountedFS));
+        if (IS_NULL_PTR(mount)) {
+            free(source);
+            free(target);
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        SMP_LOCK_INIT(mount);
+        mount->base_path = target;
+        target = NULL;
+        mount->mount_type = FATSDSPI;
+
+        ret = esp_vfs_fat_sdspi_mount(
+            mount->base_path, &host_config, &spi_slot_config, &mount_config, &mount->handle.card);
+    } else {
+        free(source);
+        free(target);
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    free(source);
+
+    term return_term = term_invalid_term();
+    if (UNLIKELY(ret != ESP_OK)) {
+        mount->mount_type = Unmounted;
+        return_term = make_esp_error_tuple(ret, ctx);
+    } else {
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2) + TERM_BOXED_RESOURCE_SIZE)
+                != MEMORY_GC_OK)) {
+            enif_release_resource(mount);
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        term mount_term = enif_make_resource(erl_nif_env_from_context(ctx), mount);
+        return_term = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(return_term, 0, OK_ATOM);
+        term_put_tuple_element(return_term, 1, mount_term);
+    }
+    enif_release_resource(mount);
```

Comment:
not working sdmmc nor sdspi.

moving `enif_release_resource(mount);` down into the umount function makes it all work.

Reply:
@petermm this surprises me, and probably @pguyot can give an hint here.

- `enif_alloc_resource` gives a resources that has already the ref counter set to 1.
- `enif_make_resource` increments it to 2
- we need `enif_make_resource` at line 273 to get it back to 1, so it can GC later.

Reply:
oh my, it appears to have been build gremlins somehow, despite everything being cleaned out:/

(changed down to otp26 and it started working, going back to otp27 kept it working :/)